editor/src/clj/editor/editor_extensions/ui_docs.clj

```diff
@@ -0,0 +1,718 @@
+(ns editor.editor-extensions.ui-docs
+  (:require [clojure.spec.alpha :as s]
+            [clojure.string :as string]
+            [editor.editor-extensions.coerce :as coerce]
+            [editor.lua-completion :as lua-completion]
+            [editor.util :as util]
+            [util.coll :as coll]))
+
+;; region make-prop
+
+;; Components are created with a table of props. The editor defines prop maps
+;; both for runtime behavior (coercion) and autocomplete/doc generation.
+
+(s/def :editor.editor-extensions.components.prop/name simple-keyword?)
+(s/def ::coerce ifn?)
+(s/def ::required boolean?)
+(s/def ::types (s/coll-of string? :min-count 1 :distinct true))
+(s/def ::doc string?)
+(s/def ::prop
+  (s/keys :req-un [:editor.editor-extensions.components.prop/name ::coerce ::required ::types ::doc]))
+
+(defn- ^{:arglists '([name & {:keys [coerce required types doc]}])} make-prop
+  "Construct a prop definition map
+
+  Args:
+    name    keyword name of a prop, should be snake_cased
+
+  Kv-args:
+    :coerce      required, coercer function
+    :required    optional, boolean, whether the prop is required to be present
+    :types       required if it can't be inferred from coercer, coll of string
+                 type names (union) for documentation
+    :doc         required, html documentation string"
+  [name & {:as m}]
+  {:post [(s/assert ::prop %)]}
+  (let [m (-> m
+              (assoc :name name)
+              (update :required boolean))]
+    (if (contains? m :types)
+      m
+      (if-let [inferred-types (condp = (:coerce m)
+                                coerce/boolean ["boolean"]
+                                coerce/string ["string"]
+                                coerce/untouched ["any"]
+                                coerce/function ["function"]
+                                nil)]
+        (assoc m :types inferred-types)
+        m))))
+
+;; endregion
+
+;; region enums
+
+;; Enums are defined separately so that the editor can define a runtime constant
+;; and documentation for each enum constant.
+
+(def enums
+  {:alignment [:top-left :top :top-right :left :center :right :bottom-left :bottom :bottom-right]
+   :padding [:small :medium :large]
+   :spacing [:small :medium :large]
+   :text_alignment [:left :center :right :justify]
+   :icon_name [:open-resource :plus :minus :clear]
+   :orientation [:vertical :horizontal]
+   :input_variant [:default :warning :error]
+   :label_variant [:default :overridden]
+   :typography_variant [:default :hint :warning :error]})
```

Comment:
I changed my mind. I'll make them kebab-cased instead. `:SCREAMING_SNAKE_CASE` keywords are hard to read in Clojure. BTW, the initial reasoning for making them snake-cased was to make `enum-prop` fn easier to use. It needs both a prop key and an enum ID, and I thought it would be nice to specify just one thing (prop name) and the enum will be automatically picked up. For example, `:text_alignment` prop has `:text_alignment` enum. But I think this is unnecessary complexity. It's no problem to specify both.